pkg/controller/devicetainteviction/device_taint_eviction.go

```diff
@@ -209,38 +238,209 @@ type taintedDevice struct {
 type allocatedClaim struct {
 	*resourceapi.ResourceClaim
 
-	// evictionTime, if non-nil, is the time at which pods using this claim need to be evicted.
+	// eviction, if non-nil, is the time at which pods using this claim need to be evicted.
 	// This is the smallest value of all such per-device values.
 	// For each device, the value is calculated as `<time of setting the taint> +
 	// <toleration seconds, 0 if not set>`.
-	evictionTime *metav1.Time
+	eviction *evictionAndReason
 }
 
-func (tc *Controller) deletePodHandler(c clientset.Interface, emitEventFunc func(tainteviction.NamespacedObject)) func(ctx context.Context, fireAt time.Time, args *tainteviction.WorkArgs) error {
-	return func(ctx context.Context, fireAt time.Time, args *tainteviction.WorkArgs) error {
-		klog.FromContext(ctx).Info("Deleting pod", "pod", args.Object)
-		var err error
-		for i := 0; i < retries; i++ {
-			err = addConditionAndDeletePod(ctx, c, args.Object, &emitEventFunc)
-			if apierrors.IsNotFound(err) {
-				// Not a problem, the work is done.
-				// But we didn't do it, so don't
-				// bump the metric.
-				return nil
-			}
-			if err == nil {
-				tc.metrics.PodDeletionsTotal.Inc()
-				tc.metrics.PodDeletionsLatency.Observe(float64(time.Since(fireAt).Seconds()))
-				return nil
-			}
-			time.Sleep(10 * time.Millisecond)
+// evictionAndReason combines the time when eviction needs to start with all reasons
+// why eviction needs to start.
+type evictionAndReason struct {
+	when   metav1.Time
+	reason evictionReason
+}
+
+func (et evictionAndReason) String() string {
+	return fmt.Sprintf("%s (%s)", et.when, et.reason)
+}
+
+func (et *evictionAndReason) equal(other *evictionAndReason) bool {
+	if (et == nil) != (other == nil) ||
+		et == nil {
+		return false
+	}
+	return et.when.Equal(&other.when) &&
+		slices.Equal(et.reason, other.reason)
+}
+
+// evictionReason collects all taints which caused eviction.
+// It supports pretty-printing for logging and inclusion in
+// user-facing descriptions
+type evictionReason []trackedTaint
+
+func (er evictionReason) String() string {
+	var parts []string
+	for _, taint := range er {
+		parts = append(parts, taint.String())
+	}
+	return strings.Join(parts, ", ")
+}
+
+// trackedTaint augments a DeviceTaint with a pointer to its origin.
+// rule and slice are mutually exclusive. At least one is set.
```

Comment:
If they are mutually exclusive why `At least one is set`? Based on the code it seems `rule` will always take priority.

Reply:
Replaced with "mutually exclusive. Exactly one of them is always set."

I wanted to make it clear that if `rule == nil`, then `slice` must be non-empty.